### PR TITLE
[popover] Add boilerplate for DOM/JS API

### PR DESCRIPTION
**Pull request by nt1m**
#### ccd8c7397ddec0fc7f0924a87b289e0ed85f137d
<pre>
[popover] Add boilerplate for DOM/JS API
<a href="https://bugs.webkit.org/show_bug.cgi?id=252077">https://bugs.webkit.org/show_bug.cgi?id=252077</a>
rdar://105388147

Reviewed by Ryosuke Niwa.

<a href="https://html.spec.whatwg.org/#the-popover-attribute">https://html.spec.whatwg.org/#the-popover-attribute</a>

- Add PopupData that stores visibility state &amp; previously focused element and in the future, invoker &amp; toggle task
- Add IDL reflection for popover attribute
- Add some basic steps for showPopover/hidePopover/togglePopover

&quot;SyntaxError: The string did not match the expected pattern.&quot; error messages are because the `:open`/`:closed` pseudo-classes aren&apos;t implemented, leading `matches(&quot;:open&quot;)` to throw.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/hide-other-popover-side-effects-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-display-none.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-nesting.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-animated-hide-cleanup.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-animation-corner-cases.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-beforetoggle-opening-event.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-document-open.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-child-dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-not-keyboard-focusable.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-shadow-dom.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-stacking.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-top-layer-combinations.tentative-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2.tentative-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss.tentative-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering.tentative-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::popoverData):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::popoverData):
(WebCore::ElementRareData::setPopoverData):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/NodeRareData.h:
* Source/WebCore/dom/PopoverData.h: Added.
(WebCore::PopoverData::visibilityState const):
(WebCore::PopoverData::setVisibilityState):
(WebCore::PopoverData::previouslyFocusedElement const):
(WebCore::PopoverData::setPreviouslyFocusedElement):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::checkPopoverValidity): Implements <a href="https://html.spec.whatwg.org/#check-popover-validity">https://html.spec.whatwg.org/#check-popover-validity</a>
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopover):
(WebCore::HTMLElement::togglePopover):
(WebCore::HTMLElement::popover const):
* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::setPopover):
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/platform/CommonAtomStrings.h:

Canonical link: <a href="https://commits.webkit.org/260218@main">https://commits.webkit.org/260218@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/23dbe24b32c395fbab246902f5920dfa29a4e944

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107556 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40458 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116716 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111451 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18039 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7929 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99724 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113315 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96813 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41282 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28439 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29792 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10276 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49373 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7062 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11837 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/html/HTMLElement.cpp [hunk 1]
```diff
@@ -1211,6 +1212,95 @@ ExceptionOr<Ref<ElementInternals>> HTMLElement::attachInternals()
     return ElementInternals::create(*this);
 }
 
+static ExceptionOr<void> CheckPopoverValidity(Element& element, PopoverVisibilityState expectedState)
```
Comment (rniwa) [inline Source/WebCore/html/HTMLElement.cpp]: Nit: it should be camelCase, not CamelCase.

File: Source/WebCore/dom/PopoverData.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Element.h"
+
+namespace WebCore {
+
+enum class PopoverVisibilityState {
```
Comment (rniwa) [inline Source/WebCore/dom/PopoverData.h]: : bool or : uint8_t?

File: Source/WebCore/dom/ElementRareData.h [hunk 3]
```diff
@@ -123,6 +124,9 @@ class ElementRareData : public NodeRareData {
 
     ExplicitlySetAttrElementsMap& explicitlySetAttrElementsMap() { return m_explicitlySetAttrElementsMap; }
 
+    PopoverData* popoverData() { return m_popoverData.get(); }
+    void setPopoverData(std::unique_ptr<PopoverData> popoverData) { m_popoverData = WTFMove(popoverData); }
```
Comment (rniwa) [inline Source/WebCore/dom/ElementRareData.h]: Missing &&.

File: Source/WebCore/html/HTMLElement.idl [hunk 4]
```diff
@@ -48,9 +48,14 @@
...
     [EnabledBySetting=FormAssociatedCustomElementsEnabled] ElementInternals attachInternals();
 
-    // Experimental: https://github.com/WICG/inert/blob/master/README.md
     [CEReactions, Reflect, EnabledBySetting=InertAttributeEnabled] attribute boolean inert;
 
+    // The popover API
+    [EnabledBySetting=PopoverAttributeEnabled] undefined showPopover();
+    [EnabledBySetting=PopoverAttributeEnabled] undefined hidePopover();
+    [EnabledBySetting=PopoverAttributeEnabled] undefined togglePopover(optional boolean force);
+    [CEReactions, EnabledBySetting=PopoverAttributeEnabled] attribute [AtomString] DOMString? popover;
```
Comment (rniwa) [inline Source/WebCore/html/HTMLElement.idl:57]: We should add a custom element reaction test. Can be done in a separate PR though.

File: Source/WebCore/platform/CommonAtomStrings.h [hunk 5]
```diff
@@ -42,6 +42,7 @@ namespace WebCore {
     macro(false, "false") \
     macro(lazy, "lazy") \
     macro(main, "main") \
+    macro(manual, "manual") \
```
Comment (rniwa) [inline Source/WebCore/platform/CommonAtomStrings.h:45]: Why is this change needed?

**Review by rniwa — APPROVED** [review]

File: Source/WebCore/platform/CommonAtomStrings.h (hunk 5 above)
Comment (nt1m) [reply Source/WebCore/platform/CommonAtomStrings.h:45]: The popover getter uses it, I expect popover="manual" to be frequently used (once popover is used more widely), so I would prefer putting it in common atom strings rather than initializing a custom atom string.

File: Source/WebCore/html/HTMLElement.idl (hunk 4 above)
Comment (nt1m) [reply Source/WebCore/html/HTMLElement.idl:57]: I'll create a separate PR to add some to LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/HTMLElement.html
